File: ruby/lib/Dice.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.  
require_relative 'GameCharacter'

module Deepspace
  class Dice
    def initialize
      @NHANGARSPROB = 0.25
      @NSHIELDSPROB = 0.25
      @NWEAPONSPROB = 0.33
      @FIRSTSHOTPROB = 0.5
      @EXTRAEFFICIENCYPROB = 0.8
      @generator = Random.new
    end 

    def initWithNHangars
      if @generator.rand <= @NHANGARSPROB
        0
      else
        1
      end
    end

    def initWithNWeapons
      #I use && instead of because of the precedence of operators
      #https://stackoverflow.com/questions/1426826/difference-between-and-and-in-ruby

      gen = @generator.rand
      if gen < @NWEAPONSPROB
        1
      elsif gen > @NWEAPONSPROB && gen <= 2*@NWEAPONSPROB
        2
      else
        3
      end
    end

    def initWithNShields
      if @generator.rand < @NSHIELDSPROB
        0
      else
        1
      end
    end

    def whoStarts(nPlayers)
      #Dos puntos es [a,b]
      #Tres puntos es [a,b)
      @generator.rand(0...nPlayers)
    end

    def firstShot
      if @generator.rand < @FIRSTSHOTPROB
        GameCharacter::SPACESTATION
      else  
        GameCharacter::ENEMYSTARSHIP
      end
    end

    def spaceStationMoves(speed)
      #We assume 0 <= speed < 1 

      if @generator.rand < speed
        true
      else
        false
      end
    end
    
    def extraEfficiency
      return @generator.rand < @EXTRAEFFICIENCYPROB
    end
  end
end

File: java/src/deepspace/Dice.java
/*
 * Here comes the text of your license
 * Each line should be prefixed with  * 
 */
package deepspace;
import java.util.Random;

/**
 *
 * @author yabir
 */
public class Dice {
    private final float NHANGARSPROB;
    private final float NSHIELDSPROB;
    private final float NWEAPONSPROB;
    private final float FIRSTSHOTPROB;
    
    Random generator;
    
    Dice(){
        this.NHANGARSPROB = 0.25f;
        this.NSHIELDSPROB = 0.25f;
        this.NWEAPONSPROB = 0.33f;
        this.FIRSTSHOTPROB = 0.5f;
        
        this.generator = new Random();
    }
    
    GameCharacter firstShot(){
        //generate a random value (double) between 0 and 1
        double random = generator.nextDouble();
        
        if(random < FIRSTSHOTPROB )
            return GameCharacter.SPACESTATION;
        else
            return GameCharacter.ENEMYSTARSHIP;
    }
    
    
    /*
    *   Number of hangars for a spacial station when 
    *   it's created
    *   @return Integers 1 or 0
    */
    public int initWithNHangars(){
        
        if (generator.nextDouble() < NHANGARSPROB)
            return 0;
        else
            return 1;
    }
    
    /*
    *   Number of weapons for a spacial station when
    *   it's created
    *   @return Integer 1,2 or 3
    */
    public int initWithNWeapons(){
        double random = generator.nextDouble();
        int result;
        
        if(random < NWEAPONSPROB)
            result = 1;
        else if (random > NWEAPONSPROB && random < 2*NWEAPONSPROB)
            result = 2;
        else
            result = 3;
        
        return result;
    }
    
    /*
    *   Number of shield's power ups that a spacial 
    *   station will receive.
    *   @return Integer 1 or 0
    */
    int initWithNShields(){
        int result = 0;
        
        if(generator.nextDouble() > NSHIELDSPROB)
            result = 1;
        
        return result;
    }
    
    /*
    *   Player that will start the game
    *   @param nPlayers Number of players
    *   @return Integer in the range [0, nPlayers-1]
    */
    
    int whoStarts(int nPlayers){
        return generator.nextInt(nPlayers); 
    }
    
    
    /*
    *   Determinate if the space station will avoid a shot.
    *   Prob of avoiding the shot will be greater if the speed is nearest the
    *   potential speed.
    *   @param A float value between 0 and 1
    */
    
    boolean spaceStationMoves(float speed){
    
        double random = generator.nextDouble();
        boolean result = true;
        
        if(random > speed)
            result = false;
        
        return result;
        
    }
}
